tracer/src/Datadog.Trace/Ci/Agent/CIAgentlessWriter.cs

```diff
@@ -243,55 +276,64 @@ private static async Task InternalFlushEventsAsync(object state)
                         buffers.FlushCiCodeCoverageBufferAsync()).ConfigureAwait(false);
 
                     // If there's a flush watermark we marked as resolved.
-                    watermarkCompletion?.TrySetResult(true);
+                    if (watermarkCountDown is not null)
+                    {
+                        watermarkCountDown.Signal();
+                        Log.Debug<int?>("CIAgentlessWriter: Waiting for signals from other buffers [Buffer: {bufferIndex}]", index);
+                        watermarkCountDown.Wait();
+                        Log.Debug<int?>("CIAgentlessWriter: Signals received, continue processing.. [Buffer: {bufferIndex}]", index);
+                    }
                 }
                 catch (ThreadAbortException ex)
                 {
                     completionSource?.TrySetException(ex);
-                    watermarkCompletion?.TrySetException(ex);
+                    watermarkCountDown?.Signal();
                     throw;
                 }
                 catch (Exception ex)
                 {
-                    watermarkCompletion?.TrySetException(ex);
+                    Log.Warning(ex, "Error in CIAgentlessWriter.InternalFlushEventsAsync");
```

Comment:
Should we have a `Signal` and `Wait` in here? 🤔

Reply:
True, done in [f849a7a](https://github.com/DataDog/dd-trace-dotnet/pull/3096/commits/f849a7ad28ab1fc594544f8127bd67fdb5dab96f)